### DISCUSSION TITLE
Limiting print output

### DISCUSSION THREAD

**Discussion by jmohdyusof** [lanl/bml#480] [Ideas]
Currently I have implemented some max_row, max_col code to limit the amount of output from tests. Does it make sense to push the limiter into the print_bml_matrix functions so they are always capped by a global constant PRINT_VARIABLE, rather than computing the threshold in every calling function?

**Comment by nicolasbock** [comment]
I think that would be helpful. It sounds like the output would be a lot less verbose that way which should make it easier to parse.

**Comment by jmohdyusof** [comment]
As per Mike Wall's comment earlier, printing out a 2000 x 2000 matrix isn't helpful. It would be nice if someone else commented to get some buy-in

**Comment by mewall** [comment]
My only comment is that it might be good to do this in a way that doesn't catch someone by surprise if they're trying to print the whole matrix and end up getting only part of it. So maybe the default is to print the whole thing and the test code decreases the limits appropriate for a log file vs. getting the matrix to record the actual data and/or to do something else with it?

________________________________
From: Jamal Mohd-Yusof ***@***.***>
Sent: Wednesday, May 5, 2021 8:25:46 AM
To: lanl/bml
Cc: Subscribed
Subject: [EXTERNAL] Re: [lanl/bml] Limiting print output (#480)


As per Mike Wall's comment earlier, printing out a 2000 x 2000 matrix isn't helpful. It would be nice if someone else commented to get some buy-in

**Comment by jmohdyusof** [comment]
So the print_matrix function already has bounds arguments build in (i_l, i_u, j_l, j_u) to allow the user to limit the range (and presumably print only local values if it is decomposed). The problem is that every time it is called it is always (0, N, 0, N). In the test codes I have modified the call, but not the function. It sounds like you are advocating keeping the function as is, and just modify all the calls in the test suite to limit output? 

Or are you saying modify the PRINT_LIMIT based on usage? Right now it is a compile-time constant set in bml_utilities.h

**Comment by jeanlucf22** [comment]
I am not sure what the goal is here. Is it because @jmohdyusof wants to use the test suite for another purpose, such as performance tests, with large matrices?

**Comment by mewall** [comment]
Yes, that is the use case we're talking about. We are using the test to obtain performance information.


I would advocate going with what was outlined in Jamal's most recent comment -- to leave the print function alone and have the tests print a smaller portion of the matrix instead of the whole matrix.

________________________________
From: Jean-Luc Fattebert ***@***.***>
Sent: Wednesday, May 5, 2021 9:17:00 AM
To: lanl/bml
Cc: Wall, Michael E; Comment
Subject: [EXTERNAL] Re: [lanl/bml] Limiting print output (#480)


I am not sure what the goal is here. Is it because @jmohdyusof<https://github.com/jmohdyusof> wants to use the test suite for another purpose, such as performance tests, with large matrices?

**Comment by jmohdyusof** [comment]
For offloaded codes, incorrect results can often happen for larger matrix sizes (> warp size) if the code is bad, so it is not just a performance test. The 'make test' doesn't include any output anyway, so what is the point of the test codes if not for this? 

In any case, it sounds like people want to be able to print arbitrary large matrices at all times, so I won't change anything

**Comment by nicolasbock** [comment]
If such a change in output is done so that one can control the output at runtime we don't have to change anything for the smaller CI tests since those matrices are all small. But then one could run with reduced output outside of CI.

**Comment by jmohdyusof** [comment]
I think we resolved this on the call: I will also note that it does appear that testf.F90 appears to hard-code N and M = 13, while bml_test.c allows N and M to be set from the command line. Should (can) these be made to both take the same arguments?

**Comment by nicolasbock** [comment]
Yes, that should be doable. Let's open an issue for that so we don't forget to change `testf.90`.

**Comment by nicolasbock** [comment]
I created #481
